go/vt/vtgate/gateway/hybridgateway.go

```diff
@@ -0,0 +1,184 @@
+/*
+Copyright 2017 Google Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gateway
+
+import (
+	"fmt"
+
+	"golang.org/x/net/context"
+
+	querypb "github.com/youtube/vitess/go/vt/proto/query"
+	topodatapb "github.com/youtube/vitess/go/vt/proto/topodata"
+	"github.com/youtube/vitess/go/vt/srvtopo"
+	"github.com/youtube/vitess/go/vt/topo"
+	"github.com/youtube/vitess/go/vt/vttablet/queryservice"
+)
+
+// HybridGateway implements the gateway.Gateway interface by forwarding
+// the queries to the right underlying implementation:
+// - it has one gateway that watches for tablets. Usually a DiscoveryGateway.
+//   Useful for local tablets, or remote tablets that can be accessed.
+// - it has a list of remote vtgate connections to talk to l2 vtgate processes.
+//   Useful for remote tablets that are far away, or if the number of local
+//   tablets grows too big.
+//
+// Note the WaitForTablets method for now only waits on the local gateway.
+type HybridGateway struct {
+	queryservice.QueryService
+
+	// gw is the local gateway that has the local connections.
+	gw Gateway
+
+	// l2vtgates is the list of remote connections to other vtgate pools.
+	l2vtgates []*L2VTGateConn
```

Comment:
Assuming HybridGateway will be built based on what's healthy, this should just be a single l2vtgate, right?
Because only the first l2 will be used.
Maybe you addressed this later. We'll see :).

Reply:
No, this is meant to be a list of l2vtgates to talk to. When using the Resolver part (and not the old routing part inside here, which I agree is bad), it will return the QueryService of the l2vtgate that has the aggregate stats for the keyspace/shard.

So after the v3 call site is converted, this hybrid gateway won't implement QueryService, but just delegate to either the underlying Gateway, or a l2vtgateconn, whichever has the stats.